### PR TITLE
Implement _finishIncompleteBlock.

### PR DESCRIPTION
**Pull request by mattcollier**
Once this gets ironed out, the same technique will be applied to events/operations.

Once this is finalized need to catch some specific errors at the consensus level.  Still not sure what our recourse is if the eventHashes fail to be saved, but I reckon we have to address that at a higher level.  Maybe it's just a matter of wiping the failed block and starting over.

Based on our current implementation with Continuity, I guess there is some application for a block that does *not* have consensus, and therefore it is left to the consensus algorithm to mark the block's `meta.consensus` etc.  Is my understanding correct?

### CODE REVIEW

File: lib/ledgerBlockStorage.js [hunk 1]
```diff
@@ -664,4 +715,63 @@ module.exports = class LedgerBlockStorage {
...
+          if(!result) {
+            return callback(new BedrockError(
+              'No failed blocks found.', 'NotFoundError'));
+          }
+          const {_status} = result.meta;
+          if(_status === this.constants.STARTED) {
+            // failure occurred before event hashes were recorded
+            return callback(new BedrockError(
+              'An unrecoverable block was found.', 'InvalidStateError',
+              result));
```
Comment (dlongley) [inline lib/ledgerBlockStorage.js]: We should never have to worry about an unrecoverable block error. Either a write goes all the way through or it "doesn't count" and we pretend it didn't happen at all. When writing a block of a particular height, we should always have all of the information we need to write it. We just don't want any of that information to be used until it is "fully written". So anyone trying to do something with a block should check the block to make sure it's status is "good" (as this is the "guard") before using any of the events that are linked to it, for example.

Here, if the block status is started, we should be able to just repeat the entire process without worrying about messing anything up. We should always have the hashes when we're trying to insert a block -- and inserting a block should only be done by a consensus process that is non-concurrent with any other consensus process. So:

1. We have the event hashes and we're trying to write a block.
2. See if we already tried and failed previously.
3. If so, just go through the whole process again.

In other words, whatever the process is for writing a block is, it should be idempotent and atomic such that if it fails at any point we can fully recover because we always have all of the information when we try to write a block. Either we complete the entire process and flip a flag that says that so we know -- or we just redo the process until we *do* finish, i.e. there should never be an "unrecoverable state" -- this is precisely what we need to avoid.

Comment (dlongley) [reply lib/ledgerBlockStorage.js]: Note: This might not be the same case for regular events. We can fail and literally lose all of the operations we were trying to store. When we fail to write a regular event, that means we literally lost the regular event and we'll need a process to clean it up. Either that, or we need to write the operations *first* (either into a temporary collection or into the actual collection, just not linked to an event yet) and then whatever we have successfully written will be what gets into the event whenever that happens (i.e. immediately or after a crash). It may be easiest to just drop stuff on the floor when we crash and consider that an acceptable risk. We'll just need a way to make sure we know when it's ok to clean up our mess.

Comment (mattcollier) [reply lib/ledgerBlockStorage.js]: The failure I'm having difficulty with here is the operation that marks events has having consensus.  I designed this patch so that if *that* update only partially completed that there is a way to recover.

The 'unrecoverable' state that you pointed out does necessarily require that consensus be recalculated, but in that case, no events from the previous failed block creation operation have been altered, so consensus can recalculate and produce the same result it had produced previously.

Comment (mattcollier) [reply lib/ledgerBlockStorage.js]: @dlongley I'm realizing that I don't know the pattern for updating many mongo records in a safe way.  The single record pattern is:
1. Set a `pending` flag
2. Make the update
3. remove the `pending` flag

So I'm getting stuck in my thinking about what happens if step 3 (removing the pending flag) fails part way through in the case of a multi-record operation.

Are you saying you want to create conditions such that if step 3 were to fail partially (let's use power outage / spurious reboot for sake of discussion), that consensus should be able to run and compute the same list of events that it had before and attempt to create the new block as if the failed attempt had never occurred?

In that situation, we would have a block that has a `pending` flag set.  Some events that have been fully tagged with `blockHeight` and `blockOrder` and had their `pending` flag removed.  And some events that have also been tagged, but have *not* had their pending flag removed.

So do we need some clean-up process that backs out all the successful tagging and remaining pending flags etc. in order to make the events eligible to run through the consensus process again?

Comment (dlongley) [reply lib/ledgerBlockStorage.js]: @mattcollier,

> Are you saying you want to create conditions such that if step 3 were to fail partially (let's use power outage / spurious reboot for sake of discussion), that consensus should be able to run and compute the same list of events that it had before and attempt to create the new block as if the failed attempt had never occurred?

Yes, consensus should be able to run and compute the same list of events as before and attempt to create the new block as if the failed attempt never occurred.

We need to ensure that there's just a single record somewhere that we can rely on for determining if some change "finished" or not. So if reading consensus events depends on first reading a consensus block, then we should have that. I may have forgotten some details here -- but it seems like we don't need any `pending` flags at all. What purpose do they serve?

Perhaps we need to step back and make sure we're clearly defining the problem.

In the case of writing a block, we need to make sure all events have been written to the database prior to writing the block. It seems like we should be able to accomplish that simply by writing the events first. If we fail at any point, we should be able to pick up where we left off.

In the case of marking a block as having achieved consensus, we need to ensure that no events get considered as having "achieved consensus" prior to the block achieving consensus. This means probably means making sure our "history" functions will pull out any events as "not having consensus" based on their `blockHeight` rather than their consensus status. IOW, we'd get the latest block that has consensus (`latestHeight`) -- and then retrieve all events that do not have consensus or that have consensus with a `blockHeight` greater than `latestHeight`. That would enable the consensus algorithm to properly recompute the events that need to go into the next block even if an update failed previously.

I don't think any of this needs `pending` flags -- but perhaps those somehow provided a more elegant solution or better indexing characteristics?

**Comment by mattcollier** [comment]
This has been re-implemented here: https://github.com/digitalbazaar/bedrock-ledger-storage-mongodb/pull/16
